### PR TITLE
[No-JIRA] Add animation when opening BpkModalV2

### PR DESCRIPTION
**Pull request by ByAnthony**
# Why has this change been made?

Following this [PR](https://github.com/Skyscanner/backpack/pull/2869), we introduced a few updates to the new version to bpk-modal-component. This PR introduces the animation when opening the BpkModalV2.

# What changed?

- Introduce animation when opening BpkModalV2;
- fix a bug when clicking on the backdrop of the polyfill, the dialog stays opened.

# Checklist

Remember to include the following changes:

- [x] Component `README.md`
- [x] Tests
- [x] Storybook examples created/updated

### CODE REVIEW

File: packages/bpk-component-modal/src/BpkModalV2/BpKModal.module.scss [hunk 1]
```diff
@@ -18,13 +18,6 @@
 
 @import '~bpk-mixins/index.scss';
 
-@mixin center-dialog {
-  top: 50%;
-  transform: translateY(-50%);
-  border-radius: $bpk-border-radius-md;
-  inset-block-start: 50%;
-}
-
```
Comment (ByAnthony) [inline packages/bpk-component-modal/src/BpkModalV2/BpKModal.module.scss:27]: Turn out the dialog element is by default center horizontally and vertically

File: packages/bpk-component-modal/src/BpkModalV2/BpKModal.module.scss [hunk 2]
```diff
@@ -50,8 +65,12 @@
 
   .bpk-modal {
     position: fixed;
+    top: 50%;
+    left: 50%;
     display: block;
     z-index: 1;
+    transform: translate(-50%, -50%) scale(1);
```
Comment (ByAnthony) [inline packages/bpk-component-modal/src/BpkModalV2/BpKModal.module.scss:72]: Still need to center as Safari 14 and 15 does support the dialog element. So need to center it

File: packages/bpk-component-modal/src/BpkModalV2/BpKModal.module.scss [hunk 3]
```diff
@@ -33,6 +26,26 @@
...
+@keyframes bpk-modal-scale {
+  0% {
+    transform: scale(0.9);
+  }
+
+  100% {
+    transform: scale(1);
+  }
+}
+
```
Comment (ByAnthony) [inline packages/bpk-component-modal/src/BpkModalV2/BpKModal.module.scss:48]: Is this the correct way to add/do animation?

File: packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx [hunk 4]
```diff
@@ -120,7 +120,7 @@ export const BpkModalV2 = (props: Props) => {
         dialog.addEventListener('click', (event: Event) => {
           const { target } = event;
 
-          if (target instanceof HTMLElement && target.id === `${id}`) {
+          if ((target as HTMLElement) && target === dialog) {
```
Comment (anambl) [inline packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx:123]: what is the reason for this change? I guess the left side will always be true as it's just casting?

Comment (ByAnthony) [reply packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx]: Oops, that was not suppose to be there

Comment (ByAnthony) [reply packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx]: Sorry, I did not pushed all my changes. I will comment on what the changes are doing.

File: packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx [hunk 5]
```diff
@@ -116,15 +116,27 @@ export const BpkModalV2 = (props: Props) => {
...
+              if (modal === dialogWithPolyfill) {
+                onClose();
+              }
+            }
+          });
+        }
+      };
+
+      handleBackdropClick(dialog);
+      handleBackdropClick(dialogWithPolyfill);
```
Comment (ByAnthony) [inline packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx:139]: Just realised I had an issue when clicking the backdrop for the polyfill. It was not working. So did a wee bit of changes to handle the click when dialog is supported or not.

File: packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx [hunk 6]
```diff
@@ -116,15 +116,22 @@ export const BpkModalV2 = (props: Props) => {
...
+
+            if (target === modal) {
+              modal === dialog ? ref.current?.close?.() : onClose();
+            }
+          });
+        }
+      };
+
+      handleBackdropClick(dialog);
+      handleBackdropClick(dialogWithPolyfill);
```
Comment (ByAnthony) [inline packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx:134]: Just realised I had an issue when clicking the backdrop for the polyfill. It was not working. So did a wee bit of changes to handle the click when dialog is supported or not.

File: packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx [hunk 7]
```diff
@@ -116,22 +116,29 @@ export const BpkModalV2 = (props: Props) => {
...
-      if (dialog) {
-        dialog.addEventListener('click', (event: Event) => {
-          const { target } = event;
-
-          if (target instanceof HTMLElement && target.id === `${id}`) {
-            ref.current?.close?.();
-          }
-        });
-      }
+      const dialogWithPolyfill = document.getElementById(`${id}-polyfill`);
```
Comment (anambl) [inline packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx:119]: is the element with `id=${id}-polyfill` not just an extra element that will be rendered along with the `dialog`?
Looking at the code it doesn't seem like it's one or the other so wondering how `modal === dialog ? ref.current?.close?.() : onClose();` this works?

Comment (anambl) [reply packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx:119]: not sure if there's a way to "hack" the dialog examples so that we can add it in Storybook 🤔

Comment (anambl) [reply packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx:119]: or maybe rephrasing the question - if `dialog` HTML element is not supported, should we still be rendering it? 🤔

Comment (anambl) [reply packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx:119]: It doesn't wrap it though - https://github.com/Skyscanner/backpack/blob/main/packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx#L156

Comment (ByAnthony) [reply packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx:119]: Sorry, I did not explain myself very well on this one. The wrapper is there on both supported/not supported. But for the unsupported it add the `id=${id}-polyfill`:  [https://github.com/Skyscanner/backpack/blob/main/packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx#L153](https://github.com/Skyscanner/backpack/blob/main/packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx#L153)

Comment (ByAnthony) [reply packages/bpk-component-modal/src/BpkModalV2/BpkModal.tsx:119]: Sorry, I got confused in my own code here. I updated my comment to reflect it.

**Review by anambl — APPROVED** [review]
